File: src/main/java/bllose/leetcode/EquiRational.java
package bllose.leetcode;

/**
 * 相等的非负有理数
 */
public class EquiRational {

    /**
     * <IntegerPart> 
     * 例： 0 ,12 和 123 
     * <IntegerPart><.><NonRepeatingPart>
     * 例： 0.5 , 1. , 2.12 和 123.0001
     * <IntegerPart><.><NonRepeatingPart><(><RepeatingPart><)> 
     * 例： 0.1(6) ， 1.(9)， 123.00(1212)
     * <p>
     * 每个部分仅由数字组成。
     * 整数部分 <IntegerPart> 不会以零开头。（零本身除外）
     * 1 <= <IntegerPart>.length <= 4
     * 0 <= <NonRepeatingPart>.length <= 4
     * 1 <= <RepeatingPart>.length <= 4
     *
     * @param s
     * @param t
     * @return
     */
    public static boolean isRationalEqualOne(String s, String t) {
        if (s.contains("(") && t.contains("(")) {
            String[] sArr = s.split("\\.");
            String[] tArr = t.split("\\.");
            if (!sArr[0].equals(tArr[0])) {
                return false;
            } else {
                String[] sDeci = sArr[1].replace(")", "").split("\\(");
                String[] tDeci = tArr[1].replace(")", "").split("\\(");

            }

        } else if (!s.contains("(") && !t.contains("(")) {
            if (Double.valueOf(s) - Double.valueOf(t) != 0) {
                return false;
            }
        } else if (s.contains("(") || t.contains("(")) {
            String bigOne;
            String smallOne;
            if (s.contains("(")) {
                bigOne = t;
                smallOne = s;
            } else {
                bigOne = s;
                smallOne = t;
            }
            String[] bigArr = bigOne.split("\\.");
            String[] smlArr = smallOne.split("\\.");

            if (bigArr.length > 1) {
                bigOne = bigArr[1];
                bigOne = bigOne.replaceAll("0", "");
                if (bigOne.length() > 0) {
                    return false;
                }
            }
            smallOne = smlArr[1];
            smallOne = smallOne.replace("(", "");
            smallOne = smallOne.replace(")", "");
            smallOne = smallOne.replaceAll("9", "");
            if (smallOne.length() > 0) {
                return false;
            }

            if (Integer.valueOf(bigArr[0]) - Integer.valueOf(smlArr[0]) != 1) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {
//        System.out.println("\"1314.1592\", \"1314.1592\" " + isRationalEqualOne("1314.1592", "1314.1592"));
//        System.out.println("\"49.324\", \"49.321\" " + isRationalEqualOne("49.324", "49.321"));
//        System.out.println("\"432.0001\", \"431.99(9)\" " + isRationalEqualOne("432.0001", "431.99(9)"));
//        System.out.println("\"432.0000\", \"431.99(9)\" " + isRationalEqualOne("432.0000", "431.99(9)"));
//        System.out.println("\"431.99(9)\", \"432.0000\" " + isRationalEqualOne("431.99(9)", "432.0000"));
//        System.out.println("\"432.\", \"431.99(9)\" " + isRationalEqualOne("432.", "431.99(9)"));
//        System.out.println("\"1.\", \"0.9999(9999)\" " + isRationalEqualOne("1.", "0.9999(9999)"));
//        System.out.println("\"1\", \"0.9999(9999)\" " + isRationalEqualOne("1", "0.9999(9999)"));

//        System.out.println("\"1314.1592\", \"1314.1592\" " + isRationalEqual("1314.1592", "1314.1592"));
//        System.out.println("\"49.324\", \"49.321\" " + isRationalEqual("49.324", "49.321"));
//        System.out.println("\"432.0001\", \"431.99(9)\" " + isRationalEqual("432.0001", "431.99(9)"));
//        System.out.println("\"432.0000\", \"431.99(9)\" " + isRationalEqual("432.0000", "431.99(9)"));
//        System.out.println("\"431.99(9)\", \"432.0000\" " + isRationalEqual("431.99(9)", "432.0000"));
//        System.out.println("\"432.\", \"431.99(9)\" " + isRationalEqual("432.", "431.99(9)"));
//        System.out.println("\"1.\", \"0.9999(9999)\" " + isRationalEqual("1.", "0.9999(9999)"));
//        System.out.println("\"1\", \"0.9999(9999)\" " + isRationalEqual("1", "0.9999(9999)"));
        System.out.println("\"0.(12)\", \"0.121(21)\" " + isRationalEqual("0.(12)", "0.121(21)"));
    }


    public static boolean isRationalEqual(String S, String T) {
        Fraction f1 = convert(S);
        Fraction f2 = convert(T);
        return f1.n == f2.n && f1.d == f2.d;
    }

    public static Fraction convert(String S) {
        int state = 0; //whole, decimal, repeating
        Fraction ans = new Fraction(0, 1);
        int decimal_size = 0;

        for (String part : S.split("[.()]")) {
            state++;
            if (part.isEmpty()) continue;
            long x = Long.valueOf(part);
            int sz = part.length();

            if (state == 1) { // whole
                ans.iadd(new Fraction(x, 1));
            } else if (state == 2) { // decimal
                ans.iadd(new Fraction(x, (long) Math.pow(10, sz)));
                decimal_size = sz;
            } else { // repeating
                long denom = (long) Math.pow(10, decimal_size);
                denom *= (long) (Math.pow(10, sz) - 1);
                ans.iadd(new Fraction(x, denom));
            }
        }
        return ans;
    }
}

class Fraction {
    long n, d;

    Fraction(long n, long d) {
        long g = gcd(n, d);
        this.n = n / g;
        this.d = d / g;
    }

    public void iadd(Fraction other) {
        long numerator = this.n * other.d + this.d * other.n;
        long denominator = this.d * other.d;
        long g = Fraction.gcd(numerator, denominator);
        this.n = numerator / g;
        this.d = denominator / g;
    }

    /**
     * Greatest Common Divisor
     */
    static long gcd(long x, long y) {
        return x != 0 ? gcd(y % x, x) : y;
    }
}

File: src/main/python/bllose/leetcode/EquiRational.py
'''
给定两个字符串 s 和 t ，每个字符串代表一个非负有理数，只有当它们表示相同的数字时才返回 true 。字符串中可以使用括号来表示有理数的重复部分。
有理数 最多可以用三个部分来表示：整数部分 、小数非重复部分 和小数重复部分 <(><)>。数字可以用以下三种方法之一来表示：
示例 1：

输入：s = "0.(52)", t = "0.5(25)"
输出：true
解释：因为 "0.(52)" 代表 0.52525252...，而 "0.5(25)" 代表 0.52525252525.....，则这两个字符串表示相同的数字。
示例 2：

输入：s = "0.1666(6)", t = "0.166(66)"
输出：true
示例 3：

输入：s = "0.9(9)", t = "1."
输出：true
解释："0.9(9)" 代表 0.999999999... 永远重复，等于 1 。[有关说明，请参阅此链接]
"1." 表示数字 1，其格式正确：(IntegerPart) = "1" 且 (NonRepeatingPart) = "" 。
提示：

每个部分仅由数字组成。
整数部分 不会以零开头。（零本身除外）
1 <= .length <= 4
0 <= .length <= 4
1 <= .length <= 4
'''
class EquiRational:
    def __init__(self, num1:str, num2:str):
        self.num1 = self.convert(num1)
        self.num2 = self.convert(num2)

    def convert(self, num:str):
        import re
        parts = re.split('[.()]', num)
        status = 0 # whole, sub, repeating
        ans = Fraction(0, 1)
        subLen = 0

        for cur in parts:
            status += 1
            if cur is None or len(cur) == 0:
                continue
            if status == 1:
                ans.add(Fraction(int(cur), 1))
            elif status == 2:
                subLen = len(cur)
                ans.add(Fraction(int(cur), 10 ** subLen))
            elif status == 3:
                denominator = (10 ** len(cur) - 1) * 10 ** subLen
                ans.add(Fraction(int(cur), denominator))
        return ans

    def compare(self) -> bool:
        return self.num1.d == self.num2.d and self.num1.n == self.num2.n


class Fraction:
    def __init__(self, numerator:int, denominator:int):
        gcd = self.greatest_common_divisor(numerator, denominator)
        self.n = int(numerator / gcd)
        self.d = int(denominator / gcd)

    def greatest_common_divisor(self, num1:int, num2:int) -> int:
        return num2 if num1 == 0 else self.greatest_common_divisor(num2%num1 , num1)

    def add(self, target) -> None:
        self.n = self.n * target.d + target.n * self.d
        self.d *= target.d
        gcd = self.greatest_common_divisor(self.n, self.d)
        self.n = int(self.n / gcd)
        self.d = int(self.d / gcd)


if __name__ == '__main__':
    # cur = Fraction(5, 25)
    # print(cur.n, cur.d)
    # addtion = Fraction(1, 10)
    # cur.add(addtion)
    # print(cur.n, cur.d)
    # print(EquiRational('0.(12)', '0.121(21)').compare())
    # print(EquiRational('0.(9)', '1').compare())
    print(EquiRational('0.9(9)', '1').compare())
